source-mongodb/backfill.go

```diff
@@ -0,0 +1,304 @@
+	// scan. Getting an estimate from the collection's metadata is very fast and should be close
+	// enough for useful logging.
+	estimatedTotalDocs, err := collection.EstimatedDocumentCount(ctx)
+	if err != nil {
+		return fmt.Errorf("getting estimated document count: %w", err)
+	}
+
+	logEntry := log.WithFields(log.Fields{
+		"database":           binding.resource.Database,
+		"collection":         binding.resource.Collection,
+		"estimatedTotalDocs": estimatedTotalDocs,
+	})
+	logEntry.Info("starting backfill for collection")
+
+	// By not specifying a sort parameter, MongoDB uses natural sort to order documents. Natural
+	// sort is approximately insertion order (but not guaranteed). We hint to MongoDB to use the _id
+	// index (an index that always exists) to speed up the process. Note that if we specify the sort
+	// explicitly by { $natural: 1 }, then the database will disregard any indices and do a full
+	// collection scan. See https://www.mongodb.com/docs/manual/reference/method/cursor.hint
+	var opts = options.Find().SetHint(bson.M{"_id": 1})
+
+	var filter = bson.D{}
+	if initialState.Backfill.LastId != nil {
+		var v interface{}
+		if err := initialState.Backfill.LastId.Unmarshal(&v); err != nil {
+			return fmt.Errorf("unmarshalling last_id: %w", err)
+		}
+		filter = bson.D{{idProperty, bson.D{{"$gt", v}}}}
+	}
+
+	cursor, err := collection.Find(ctx, filter, opts)
+	if err != nil {
+		return fmt.Errorf("collection.Find: %w", err)
+	}
+	defer cursor.Close(ctx)
+
+	// docs represents a batch of serialized documents read from the cursor which are ready to be
+	// output.
+	var docs []json.RawMessage
+	batchBytes := 0
+
+	var lastId bson.RawValue
+
+	docCount := 0
+	totalDocsCaptured := initialDocCount
+
+	// emitDocs is a helper for emitted a batch of documents.
+	emitDocs := func() error {
+		checkpoint := captureState{
+			Resources: map[boilerplate.StateKey]resourceState{
+				sk: {Backfill: backfillState{LastId: &lastId, BackfilledDocs: totalDocsCaptured}},
+			},
+		}
+
+		if cpJson, err := json.Marshal(checkpoint); err != nil {
+			return fmt.Errorf("serializing checkpoint: %w", err)
+		} else if err := c.output.DocumentsAndCheckpoint(cpJson, true, binding.index, docs...); err != nil {
+			return fmt.Errorf("ouutputting documents and checkpoint: %w", err)
+		}
+
+		// Reset for the next batch.
+		docs = nil
+		batchBytes = 0
+
+		c.mu.Lock()
+		state := c.state.Resources[sk]
+		state.Backfill.LastId = &lastId
+		state.Backfill.BackfilledDocs = totalDocsCaptured
+		c.state.Resources[sk] = state
+		c.mu.Unlock()
+
+		return nil
+	}
+
+	for cursor.Next(ctx) {
+		docCount++
+		totalDocsCaptured++
+
+		var doc bson.M
+		if lastId, err = cursor.Current.LookupErr(idProperty); err != nil {
+			return fmt.Errorf("looking up idProperty: %w", err)
+		} else if err = cursor.Decode(&doc); err != nil {
+			return fmt.Errorf("backfill decoding document: %w", err)
+		}
```

Comment:
Is there a reason to prefer `LookupErr(idProperty)` over `doc[idProperty]`?

Reply:
It's a really nice way to get a `bson.RawValue` that works for checkpointing. We could get a `primitive.Binary` by `doc[idProperty]` and I'm not sure what we'd have to do with that to make sure it serializes and deserializes from a checkpoint JSON. We used to decode twice; first into a `map[string]bson.RawValue` to get the `bson.RawValue` of the id, and then into a `bson.M`, see [here](https://github.com/estuary/connectors/blob/3c0b7d31744f19af05359ad9326b62f6ddca208e/source-mongodb/pull.go#L684-L688), seemingly getting the `bson.RawValue` for good reason.  Using `Lookup` is much faster.

`Lookup` traverses into nested fields if you provide more than one key. So `Lookup("some", "nested")` would find `"value"` in `{"some": {"nested": "value"}}` as a simplified example. `LookupErr` will return an error if the `idProperty` is not present in the top-level of the document, which is where we need to find it.

I ran this as a sanity check:
```go
func TestLookupErr(t *testing.T) {
	val := bson.D{{"some", bson.D{{"nested", "value"}}}}

	b, err := bson.Marshal(val)
	require.NoError(t, err)

	raw := bson.Raw(b)

	_, err = raw.LookupErr("nested")
	require.Error(t, err)

	got, err := raw.LookupErr("some")
	require.NoError(t, err)
	fmt.Println(got) // {"nested": "value"}

	got, err = raw.LookupErr("some", "nested")
	require.NoError(t, err)
	fmt.Println(got) // {"value"}
}
```